eventlog/eventlog.py

```diff
@@ -0,0 +1,759 @@
+    def validate(self) -> Tuple[bool, bool, str]:
+        return True, True, ''
+
+# ########################################
+# Boolean variable readout event
+# ########################################
+
+class EfiVarBooleanEvent(EfiVarEvent):
+    def __init__ (self, eventheader: Tuple, buffer: bytes, idx: int):
+        super().__init__(eventheader, buffer, idx)
+        (self.enabled,) =  struct.unpack('<?', self.data[:1])
+
+    def toJson (self) -> dict:
+        j = super().toJson()
+        j['Event']['VariableData'] = { 'Enabled' : 'Yes' if self.enabled else 'No' }
+        return j
+
+# ########################################
+# String event
+# ########################################
+
+class EfiVarStringEvent(EfiVarEvent):
+    def toJson (self) -> dict:
+        j = super().toJson()
+        j['Event']['VariableData'] = { 'String' : self.data.decode('utf-8') }
+        return j
+
+# ########################################
+# Hex event
+# ########################################
+
+class EfiVarHexEvent(EfiVarEvent):
+    def toJson (self) -> dict:
+        j = super().toJson()
+        j['Event']['VariableData'] = self.data.hex()
+        return j
+
+# ########################################
+# EFI variable: boot entry
+# ########################################
+
+class EfiVarBootEvent (EfiVarEvent):
+    def __init__ (self, eventheader: Tuple, buffer: bytes, idx: int):
+        super().__init__(eventheader, buffer, idx)
+        # EFI_LOAD_OPTION, https://dox.ipxe.org/UefiSpec_8h_source.html, line 2069
+        (self.attributes, self.filepathlistlength) = struct.unpack('<IH', self.data[0:6])
+        # description UTF-16 string: from byte 6 to the first pair of zeroes
+        desclen = 0
+        while self.data[desclen+6:desclen+8] != bytes([0,0]):
+            desclen += 2
+        self.description = self.data[6:6+desclen]
+        # dev path: from the end of the description string to the end of data
+        devpathlen = (self.datalen - 8 - desclen) * 2 + 1
+#        try:
+#            self.devicePath = efiDevicePath(self.data[8+desclen:8+desclen+devpathlen], devpathlen)
+#        except Exception as _:
+        self.devicePath = self.data[8+desclen:8+desclen+devpathlen].hex()
+
+    @classmethod
+    def Parse(cls, eventheader: Tuple, buffer: bytes, idx: int):
+        (namelen,) = struct.unpack('<Q', buffer[idx+16:idx+24])
+        name = buffer[idx+32:idx+32+2*namelen].decode('utf-16')
+        if name == 'BootOrder':
+            return EfiVarBootOrderEvent(eventheader, buffer, idx)
+        if re.compile('^Boot[0-9a-fA-F]{4}$').search(name):
+            return EfiVarBootEvent (eventheader, buffer, idx)
+        return EfiVarEvent(eventheader, buffer, idx)
+
+    # the published digest can match the entire event buffer or just the data portion (without the name).
+    def validate(self) -> Tuple[bool,bool,str]:
+        for algid in self.digests.keys():
+            refdigest = self.digests[algid].digest
```

Comment:
Use the iteration pattern that I commented before

Reply:
done